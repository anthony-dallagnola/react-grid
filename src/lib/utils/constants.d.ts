declare const ERRORS: {
    HEADER: {
        UNDEFINED_OR_NULL: string;
        EMPTY: string;
        MIXING_SINGLE_AND_DOUBLE: string;
        INVALID_FORMAT: string;
    };
    DATA: {
        UNDEFINED_OR_NULL: string;
        EMPTY: string;
        LENGTH: string;
        INVALID_TYPE_CELL: string;
        MISSING_VALUE_FIELD: string;
    };
    WEIGHTS: {
        LENGTH: string;
        INVALID_FORMAT: string;
    };
};
declare const PREFIX = "react-cell-grid";
export { ERRORS, PREFIX, };
