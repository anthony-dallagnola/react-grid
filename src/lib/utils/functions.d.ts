/**
 * Checks whether the variable is primitive or not
 *
 * @param {any} value value we want to check
 * @returns {boolean} true if variable is primitive false otherwise
 */
declare function isPrimitive(value: any): boolean;
export { isPrimitive, };
