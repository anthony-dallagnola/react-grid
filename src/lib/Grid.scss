$grid-border-size: 2px;
$grid-border-color: black;

.react-cell-grid-container {
  margin: auto;
  border: $grid-border-size $grid-border-color solid;
}

.react-cell-grid-container .react-cell-grid-row div, .react-cell-grid-container .react-cell-grid-header div {
  border-right: $grid-border-size $grid-border-color solid;
}

.react-cell-grid-container .react-cell-grid-row div:last-child, .react-cell-grid-container .react-cell-grid-header div:last-child {
  border: 0;
}

.react-cell-grid-row, .react-cell-grid-header {
  display: flex;
  border-bottom: $grid-border-size $grid-border-color solid;
}

.react-cell-grid-header-cell div:first-child {
  border: 0;
}

.react-cell-grid-row:last-child {
  border: 0;
}

.react-cell-grid-row:nth-child(2n) {
    background-color: lightgray;
}

.react-cell-grid-header-cell {
  text-align: center;
  position: relative;
  line-height: 20px;
}

/* sort icons */
.react-cell-grid-sort-icon-container img {
  height: 20px;
}

.react-cell-grid-sort-icon {
  height: 15px;
  position: absolute;
  right: 0;
  top: 2.5px;
  cursor: pointer;
}

.react-cell-grid-cell {
  font-size: 12px;
  display: inline-block;
}

.react-cell-grid-data-cell div {
  margin: 2px 5px;
  word-wrap: break-word;
}