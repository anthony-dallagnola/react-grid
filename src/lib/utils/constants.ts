const ERRORS = {
  HEADER: {
    UNDEFINED_OR_NULL: 'You haven\'t sent any header',
    EMPTY: 'You have sent empty headers',
    MIXING_SINGLE_AND_DOUBLE: 'You can\'t mix single and double row headers',
    INVALID_FORMAT: 'Format of the header is invalid',
  },
  DATA: {
    UNDEFINED_OR_NULL: 'You haven\'t sent any data',
    EMPTY: 'You have sent empty data',
    LENGTH: 'Problem with the data length, it doesn\'t match the headers',
    INVALID_TYPE_CELL: 'The data cell is neither a Primitive nor an object',
    MISSING_VALUE_FIELD: 'The data cell is an object but missing `value` field',
  },
  WEIGHTS: {
    LENGTH: 'Problem with the weights length, it doesn\'t match the headers',
    INVALID_FORMAT: 'The weights has invalid format, it should be an array of numbers (int or float)',
  },
  DEFAULT_SORT: {
    INVALID_FORMAT: 'The default sort props has an invalid format',
    INEXISTANT_HEADER: 'The default sort props is based on an inexistant header',
  },
};

const PREFIX = 'react-cell-grid';

export {
  ERRORS,
  PREFIX,
};
