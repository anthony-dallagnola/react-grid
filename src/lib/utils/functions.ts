/**
 * Checks whether the variable is primitive or not
 *
 * @param {any} value value we want to check
 * @returns {boolean} true if variable is primitive false otherwise
 */
function isPrimitive(value: any): boolean {
  return ['string', 'number', 'boolean'].includes(typeof value);
}

export {
  isPrimitive,
};
