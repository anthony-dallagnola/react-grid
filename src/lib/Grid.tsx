import React, {useState, useEffect} from 'react';
import { isPrimitive } from './utils/functions';
import { ERRORS, PREFIX } from './utils/constants';
import { Primitive } from './models/Primitive';
import sortUpIcon from './assets/sort-up.svg';
import sortDownIcon from './assets/sort-down.svg';
import sortBothIcon from './assets/sort-both.svg';
import { Sort } from './models/enums/Sort.enum';
import { DataCell } from './models/DataCell';
import { SortIcon, GridContainer, DivHeaderCell, GridHeader, GridRow, DivDataCell, DivDataCellContent } from './Grid.style';
import { GridProps } from './models/GridProps';
import { UserProps } from './models/UserProps';

/**
 * Type guard function
 *
 * @param {Primitive | DataCell} cell cell
 * @returns {boolean} true if is DataCell
 */
function isDataCell(cell: Primitive | DataCell): cell is DataCell {
  return (cell as DataCell).value !== undefined;
}

/**
 * Extracts value for comparing
 * If it's a string it applies lowerCase
 *
 * @param {Primitive | DataCell} cell cell data
 * @returns {Primitive} value fo the cell
 */
const extractValue = (cell: Primitive | DataCell): Primitive => {
  if (isDataCell(cell)) {
    return typeof cell.value === 'string' ? cell.value.toLowerCase() : cell.value;
  } else {
    return typeof cell === 'string' ? cell.toLowerCase() : cell;
  }
};


/**
 * Initializes grid variable from props
 *
 * @param {UserProps} props props
 * @returns {GridProps} props
 */
function init(props: UserProps): GridProps {
  return {
    intitialized: false,
    headers: props.headers,
    isSingleRowHeader: props.headers ? props.headers.filter((e) => isPrimitive(e)).length > 0 : false,
    data: props.data,
    weights: props.weights || [],
    weightsSum: 0,
    sort: props.defaultSort || {},
  };
}

let logger: any = {};
const Grid = (props: any): JSX.Element => {
  // disable logging if debug is not set
  if (props.debug) {
    logger = {
      log: (...args: any[]) => console.log(...[PREFIX+': '].concat(args)),
      warn: (...args: any[]) => console.warn(...[PREFIX+': '].concat(args)),
    };
  } else {
    logger = {
      log: () => { },
      warn: () => { },
    };
  }


  const [grid, setGrid] = useState(init(props));

  useEffect(() => {
    sortData();
  }, [grid.sort]);

  useEffect(() => {
    setGrid(init(props));
  }, [props]);

  logger.log('Data init: ', grid.data);
  logger.log('Data init: ', props.data);


  const sortData = () => {
    logger.log('sort data: ', grid.sort);
    if (typeof grid.sort === 'undefined' || typeof grid.sort.name === 'undefined' || typeof grid.sort.type === 'undefined') {
      return;
    }
    // get column number
    let index = 0;
    if (grid.isSingleRowHeader) {
      for (let i = 0; i < grid.headers.length; i++) {
        if (grid.headers[i] === grid.sort.name) {
          break;
        }
        index++;
      }
    } else {
      let found = false;
      for (let i = 0; i < grid.headers.length; i++) {
        for (let j = 0; j < grid.headers[i][1].length; j++) {
          if (grid.headers[i][1][j] === grid.sort.name) {
            found = true;
            break;
          }
          index++;
        }
        if (found) break;
      }
    }
    logger.log('column: ', index);
    // TODO check if index exceeds length
    let data: any[][];
    if (grid.sort.type === Sort.Both) {
      data = [].concat(props.data);
    } else {
      data = [].concat(props.data).sort((a: Primitive[] | DataCell[], b) => {
        let result: number;
        if (grid.sort.type === Sort.Asc) {
          result = extractValue(a[index]) > extractValue(b[index]) ? 1 : -1;
        } else {
          result = extractValue(b[index]) > extractValue(a[index]) ? 1 : -1;
        }
        return result;
      });
    }
    setGrid((oldGrid) => ({
      ...oldGrid,
      data,
    }));
  };

  const getWidthFromIndex = (index: number) => {
    return getWidth(grid.weights[index]);
  };

  const getWidth = (weight: number) => {
    return 'calc(' + (weight / grid.weightsSum * 100) + '% - 2px)';
  };

  logger.log('sort: ', grid.sort);
  const onChangeSort = (name: Primitive) => {
    if (grid.sort.name === name) {
      setGrid((oldGrid: GridProps) => ({
        ...oldGrid,
        sort: {
          ...oldGrid.sort,
          type: ((oldGrid.sort.type+1)%3),
        },
      }));
    } else {
      logger.log('ASC');
      setGrid((oldGrid: GridProps) => ({
        ...oldGrid,
        sort: {
          name,
          type: Sort.Asc,
        },
      }));
      // setSort({name, type: Sort.Asc});
    }
  };

  const getSortIcons = (name: Primitive) => {
    if (!props.sort) {
      return <></>;
    } else if (!grid.sort || grid.sort.name !== name || grid.sort.type === Sort.Both) {
      return <SortIcon onClick={() => onChangeSort(name)} src={sortBothIcon}/>;
    } else if (grid.sort.type === Sort.Asc) {
      return <SortIcon onClick={() => onChangeSort(name)} src={sortUpIcon}/>;
    } else {
      return <SortIcon onClick={() => onChangeSort(name)} src={sortDownIcon}/>;
    }
  };

  if (!grid.intitialized) {
    logger.log('headers: ', grid.headers);
    // check number of columns
    if (grid.headers == null) {
      return <div>{ERRORS.HEADER.UNDEFINED_OR_NULL}</div>;
    }
    if (grid.headers.length === 0) {
      return <div>{ERRORS.HEADER.EMPTY}</div>;
    }
    let numberColumns = 0;
    // let isSingleRowHeader = null;

    for (const header of grid.headers) {
      const isHeaderPrimitive = isPrimitive(header);
      if (!isHeaderPrimitive && !Array.isArray(header)) {
        return <div>{ERRORS.HEADER.INVALID_FORMAT}</div>;
      }
      if (grid.isSingleRowHeader !== isHeaderPrimitive) {
        return <div>{ERRORS.HEADER.MIXING_SINGLE_AND_DOUBLE}</div>;
      }
      if (isHeaderPrimitive) {
        numberColumns++;
      } else if (Array.isArray(header) && header.length === 2 && isPrimitive(header[0]) && Array.isArray(header[1])) {
        numberColumns += header[1].length;
      } else {
        return <div>{ERRORS.HEADER.INVALID_FORMAT}</div>;
      }
    }
    logger.log('numberColumns: ', numberColumns);

    logger.log('data: ', grid.data);

    if (grid.data == null) {
      return <div>{ERRORS.DATA.UNDEFINED_OR_NULL}</div>;
    }
    if (grid.data.length === 0) {
      return <div>{ERRORS.DATA.EMPTY}</div>;
    }

    for (const row of grid.data) {
      if (row.length !== numberColumns) {
        return <div>{ERRORS.DATA.LENGTH}</div>;
      }
      for (const cell of row) {
        if (isPrimitive(cell)) {

        } else if (typeof cell === 'object') {
          if (Array.isArray(cell)) {
            return <div>{ERRORS.DATA.INVALID_TYPE_CELL}</div>;
          } else if (!(cell as Object).hasOwnProperty('value')) {
            return <div>{ERRORS.DATA.MISSING_VALUE_FIELD}</div>;
          }
        }
      }
    }
    // headers.length is wrong for double header
    let weights = grid.weights;
    if (weights.length === 0) {
      weights = Array(numberColumns).fill(0, 0).map(() => 1);
    } else if (weights.length !== numberColumns) {
      return <div>{ERRORS.WEIGHTS.LENGTH}</div>;
    } else {
      for (const weight of weights) {
        if (typeof weight !== 'number') {
          return <div>{ERRORS.WEIGHTS.INVALID_FORMAT}</div>;
        }
      }
    }
    logger.log('weights: ', weights);
    const weightsSum = weights.reduce((total: any, value: any) => total + value);
    logger.log('weightsSum: ', weightsSum);

    if (props.sort) {
      if (typeof props.defaultSort !== 'undefined') {
        if (typeof props.defaultSort !== 'object' || typeof props.defaultSort.name === 'undefined' || !isPrimitive(props.defaultSort.name) ||
         typeof props.defaultSort.type === 'undefined' || ![Sort.Both, Sort.Asc, Sort.Desc].includes(parseInt(grid.sort.type))) {
          return <div>{ERRORS.DEFAULT_SORT.INVALID_FORMAT}</div>;
        }
        if (grid.isSingleRowHeader) {
          if (!grid.headers.includes(props.defaultSort.name)) {
            return <div>{ERRORS.DEFAULT_SORT.INEXISTANT_HEADER}</div>;
          }
        } else if (!grid.headers.reduce((t: Primitive[], v: Primitive[][]) => t.concat(v[1]), []).includes(props.defaultSort.name)) {
          return <div>{ERRORS.DEFAULT_SORT.INEXISTANT_HEADER}</div>;
        }
      }
    } else if (props.defaultSort) {
      logger.warn('sortDefault is set but sort is not enabled with sort={true}');
    }
    setGrid((oldGrid) => ({
      ...oldGrid,
      intitialized: true,
      weights,
      weightsSum,
    }));
  }

  return (
    <>
      <GridContainer>
        {/* Header(s) */}
        {(() => {
          const headersRow = [];
          let result;
          if (Array.isArray(grid.headers[0])) {
            // we have double header format
            let weightIndex = 0;
            const topHeadersRow = [];
            const headersRowTemp = [];
            for (let i = 0; i < grid.headers.length; i++) {
              let weight = 0;
              for (let j = 0; j < grid.headers[i][1].length; j++) {
                headersRowTemp.push(
                  <DivHeaderCell key={grid.headers[i][1][j]+'_'+i+'_'+j} width={getWidthFromIndex(weightIndex)}>
                    <div>{grid.headers[i][1][j]}</div>
                    {getSortIcons(grid.headers[i][1][j])}
                  </DivHeaderCell>);
                weight += grid.weights[weightIndex++];
              }
              topHeadersRow.push(
                <DivHeaderCell key={grid.headers[i][0]+'_'+i} width={getWidth(weight)}>{grid.headers[i][0]}
                </DivHeaderCell>);
            }
            result = <>
              <GridHeader key={'header_0'} >{topHeadersRow}</GridHeader>
              <GridHeader key={'header_1'} >{headersRowTemp}</GridHeader>
            </>;
          } else {
            // single header
            for (let i = 0; i < grid.headers.length; i++) {
              headersRow.push(<DivHeaderCell key={grid.headers[i]} width={getWidthFromIndex(i)}>
                <div>{grid.headers[i]}</div>
                {/* <GetSortIcons name={headers[i]}/> */}
                {getSortIcons(grid.headers[i])}
              </DivHeaderCell>);
            }
            result = <GridHeader >{headersRow}</GridHeader>;
          }
          return result;
        })()}
        {/* Rows */}
        <div>
          {(() => {
            const dataRows = [];
            for (let i = 0; i < grid.data.length; i++) {
              // logger.log('data rendering [', i, ']: ', data[i]);
              dataRows.push(<GridRow key={'row_'+i} >{(() => {
                const dataRow = [];
                for (let j = 0; j < grid.data[i].length; j++) {
                  if (typeof grid.data[i][j] === 'object') {
                    dataRow.push(<DivDataCell key={grid.data[i][j].value+'_'+i+'_'+j} width={getWidthFromIndex(j)}>
                      <a href={grid.data[i][j]._href}><DivDataCellContent>{grid.data[i][j].value}</DivDataCellContent></a>
                    </DivDataCell>);
                  } else {
                    dataRow.push(
                      <DivDataCell key={grid.data[i][j]+'_'+i+'_'+j} width={getWidthFromIndex(j)}>
                        <DivDataCellContent>{grid.data[i][j]}</DivDataCellContent>
                      </DivDataCell>);
                  }
                }
                return dataRow;
              })()}</GridRow>);
            }
            return dataRows;
          })()}
        </div>
      </GridContainer>
    </>
  );
};

export default Grid;
