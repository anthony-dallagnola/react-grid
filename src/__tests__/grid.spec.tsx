import React from 'react';
import { shallow, HTMLAttributes, mount, ReactWrapper } from 'enzyme';
import Grid from '../lib/Grid';
import { ERRORS } from '../lib/utils/constants';
import renderer from 'react-test-renderer';
import '../setupTests';

const propsHeader: any = {
  // debug: true,
};

describe('Header:', () => {
  it('Undefined or null', () => {
    testExpectError({ ...propsHeader }, ERRORS.HEADER.UNDEFINED_OR_NULL);
    testExpectError({ ...propsHeader, headers: null }, ERRORS.HEADER.UNDEFINED_OR_NULL);
  });

  it('Empty', () => {
    testExpectError({ ...propsHeader, headers: [] }, ERRORS.HEADER.EMPTY);
  });

  it('Mixing single and double header', () => {
    testExpectError({ ...propsHeader, headers: ['Hello', ['There', ['and there']]] }, ERRORS.HEADER.MIXING_SINGLE_AND_DOUBLE);
  });

  it('Invalid format', () => {
    /* single row headers */
    testExpectError({ ...propsHeader, headers: ['Hello', {}] }, ERRORS.HEADER.INVALID_FORMAT);

    /* double row headers */
    // object instead of primitive
    testExpectError({ ...propsHeader, headers: [['Hello', ['']], { a: ['There', ['and there']] }] }, ERRORS.HEADER.INVALID_FORMAT);
    // too many elements
    testExpectError({ ...propsHeader, headers: [['Hello', ['']], ['There', ['and there'], 'also there']] }, ERRORS.HEADER.INVALID_FORMAT);
    // first element of titles is an object
    testExpectError({ ...propsHeader, headers: [['Hello', ['']], [{}, ['and there']]] }, ERRORS.HEADER.INVALID_FORMAT);
    // list of titles is a primitive
    testExpectError({ ...propsHeader, headers: [['Hello', ['']], ['There', 'and there']] }, ERRORS.HEADER.INVALID_FORMAT);
  });
});
// we add a valid headers parameter
const propsData = {
  ...propsHeader,
  headers: ['a', 'b', 'c'],
};
describe('Data: ', () => {
  it('Undefined or null', () => {
    testExpectError({ ...propsData }, ERRORS.DATA.UNDEFINED_OR_NULL);
    testExpectError({ ...propsData, data: null }, ERRORS.DATA.UNDEFINED_OR_NULL);
  });
  it('Empty', () => {
    testExpectError({ ...propsData, data: [] }, ERRORS.DATA.EMPTY);
  });
  it('Rows length mismatch header', () => {
    testExpectError({ ...propsData, data: [['a', 'b']] }, ERRORS.DATA.LENGTH);
    testExpectError({ ...propsData, data: [['a', 'b', 'c'], ['a', 'b']] }, ERRORS.DATA.LENGTH);
  });
  it('Data cell invalid format', () => {
    testExpectError({ ...propsData, data: [['a', 'b', []]] }, ERRORS.DATA.INVALID_TYPE_CELL);
  });
  it('Data cell missing value field', () => {
    testExpectError({ ...propsData, data: [['a', 'b', {}]] }, ERRORS.DATA.MISSING_VALUE_FIELD);
  });
});

const propsWeights = {
  ...propsData,
  data: [[1, 2, 3]],
};

describe('Weights: ', () => {
  it('Length', () => {
    testExpectError({ ...propsWeights, weights: [1.2, 3] }, ERRORS.WEIGHTS.LENGTH);
  });
  it('Invalid format', () => {
    testExpectError({ ...propsWeights, weights: ['a', 1, 2] }, ERRORS.WEIGHTS.INVALID_FORMAT);
  });
});


const propsSort = {
  ...propsWeights,
  sort: true,
};
describe('Sort: ', () => {
  it('Icons not created', () => {
    const grid = shallow(<Grid {...propsSort} sort={false} />);
    const images = grid.find('SortIcon');
    expect(images).toHaveLength(0);
  });

  it('Icons ok', () => {
    const grid = mount(<Grid {...propsSort} />);
    let images = grid.find('SortIcon');
    // default sort
    expect(images).toHaveLength(3);
    for (let i = 0; i < images.length; i++) {
      expect(images.at(i).props().src).toBe('sort-both.svg');
    }
    images.at(1).simulate('click');
    // sort up
    images = grid.find('SortIcon');
    testExpectSrcFromArray(images, ['sort-both.svg', 'sort-up.svg', 'sort-both.svg']);

    images.at(1).simulate('click');
    // sort down
    images = grid.find('SortIcon');
    testExpectSrcFromArray(images, ['sort-both.svg', 'sort-down.svg', 'sort-both.svg']);
    images.at(1).simulate('click');
    // back to default sort
    images = grid.find('SortIcon');
    expect(images).toHaveLength(3);
    for (let i = 0; i < images.length; i++) {
      expect(images.at(i).props().src).toBe('sort-both.svg');
    }
    // click on third column
    images.at(1).simulate('click');
    images.at(2).simulate('click');
    images = grid.find('SortIcon');
    testExpectSrcFromArray(images, ['sort-both.svg', 'sort-both.svg', 'sort-up.svg']);
  });

  it('Default sort ok', () => {
    const defaultSort = {name: 'b', type: 1};
    const grid = mount(<Grid {...propsSort} defaultSort={defaultSort} />);
    const images = grid.find('SortIcon');
    testExpectSrcFromArray(images, ['sort-both.svg', 'sort-up.svg', 'sort-both.svg']);
  });

  it('Default sort invalid format', () => {
    testExpectError({...propsSort, defaultSort: {}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: []}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: 'a'}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {type: 0}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: ['a'], type: 0}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: {a: 'a'}, type: 0}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: null, type: 0}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: 'a', type: 'a'}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: 'a', type: ['a']}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: 'a', type: {a: 'a'}}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
    testExpectError({...propsSort, defaultSort: {name: 'a', type: null}}, ERRORS.DEFAULT_SORT.INVALID_FORMAT);
  });

  it('Default sort on inexisting header', () => {
    testExpectError({ ...propsSort, defaultSort: { name: 'd', type: 0 } }, ERRORS.DEFAULT_SORT.INEXISTANT_HEADER);
    testExpectError({ ...propsSort, headers: [['t', ['a', 'b']], ['t2', ['c']]], defaultSort: { name: 'd', type: 0 } }, ERRORS.DEFAULT_SORT.INEXISTANT_HEADER);
  });
});

describe('Snapshot on Grid', () => {
  it('renders correctly', () => {
    const rendered = renderer.create(
      <Grid {...propsSort} />,
    );
    expect(rendered.toJSON()).toMatchSnapshot();
  });
});

/**
 * Tests simple errors that render div with error as text
 *
 * @param {Object} props object containing the props to build the Grid component
 * @param {String} error expected error
 * @returns {Void} void
 */
function testExpectError(props: object, error: string): void {
  const grid = shallow(<Grid {...props} />);
  const div = grid.find('div');
  expect(div).toHaveLength(1);
  expect(grid.find('div').text()).toBe(error);
}

/**
 * Tests if all the img have the right src value
 *
 * @param {ReactWrapper} images list of image markups
 * @param {String[]} expectedList list of expected srcs
 * @returns {Void} void
 */
function testExpectSrcFromArray(images: ReactWrapper<HTMLAttributes, any, React.Component<{}, {}, any>>, expectedList: string[]): void {
  expect(images).toHaveLength(expectedList.length);
  for (let i = 0; i < expectedList.length; i++) {
    expect(images.at(i).props().src).toBe(expectedList[i]);
  }
}
