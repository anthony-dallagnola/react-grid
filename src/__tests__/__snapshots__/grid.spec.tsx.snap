// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot on Grid renders correctly 1`] = `
<div
  className="sc-AxjAm nLOwl"
>
  <div
    className="sc-AxirZ celHmN"
  >
    <div
      className="sc-AxhUy liDVmT"
      width="calc(33.33333333333333% - 2px)"
    >
      <div>
        a
      </div>
      <img
        className="sc-AxhCb oFDjl"
        onClick={[Function]}
        src="sort-both.svg"
      />
    </div>
    <div
      className="sc-AxhUy liDVmT"
      width="calc(33.33333333333333% - 2px)"
    >
      <div>
        b
      </div>
      <img
        className="sc-AxhCb oFDjl"
        onClick={[Function]}
        src="sort-both.svg"
      />
    </div>
    <div
      className="sc-AxhUy liDVmT"
      width="calc(33.33333333333333% - 2px)"
    >
      <div>
        c
      </div>
      <img
        className="sc-AxhCb oFDjl"
        onClick={[Function]}
        src="sort-both.svg"
      />
    </div>
  </div>
  <div>
    <div
      className="sc-AxiKw gdJtAp"
    >
      <div
        className="sc-AxgMl fAoDLv"
        width="calc(33.33333333333333% - 2px)"
      >
        <div
          className="sc-AxheI kfcUCx"
        >
          1
        </div>
      </div>
      <div
        className="sc-AxgMl fAoDLv"
        width="calc(33.33333333333333% - 2px)"
      >
        <div
          className="sc-AxheI kfcUCx"
        >
          2
        </div>
      </div>
      <div
        className="sc-AxgMl fAoDLv"
        width="calc(33.33333333333333% - 2px)"
      >
        <div
          className="sc-AxheI kfcUCx"
        >
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;
